File: tools/agent-tries-baml/services/api/routers/workers.py
"""Worker presence endpoints for the dashboard's live agents roster.

Every long-lived processor POSTs a heartbeat here (~15s) via
bench_core.processor; the UI reads GET /workers through /api/state.
Presence is observability only — never load-bearing for queue correctness.
"""

from __future__ import annotations

from typing import Any, Optional

from fastapi import APIRouter, Response
from pydantic import BaseModel

from ..convex_gateway import ConvexGateway


class HeartbeatBody(BaseModel):
    """Request body for a worker presence heartbeat."""

    workerId: str
    role: str
    status: str  # idle | busy
    currentItemId: Optional[str] = None


def make_workers_router(convex: ConvexGateway) -> APIRouter:
    """Build the worker-presence router.

    Args:
        convex: Gateway used to invoke the workers Convex functions.

    Returns:
        An APIRouter exposing GET /workers and POST /workers/heartbeat.
    """
    r = APIRouter(prefix="/workers", tags=["workers"])

    @r.get("")
    async def list_(role: Optional[str] = None, limit: int = 100) -> list[dict[str, Any]]:
        """List worker presence rows, optionally filtered by role.

        Args:
            role: Optional role to filter on.
            limit: Maximum number of rows to return.

        Returns:
            The matching worker documents.
        """
        args: dict[str, Any] = {"limit": limit}
        if role is not None:
            args["role"] = role
        return await convex.query("workers:list", args)

    @r.post("/heartbeat")
    async def heartbeat(body: HeartbeatBody) -> Response:
        """Upsert a worker's presence row, stamping lastHeartbeat to now.

        Args:
            body: The worker's identity, role, status, and current item.

        Returns:
            An empty 204 Response.
        """
        await convex.mutation("workers:upsert", body.model_dump(exclude_none=True))
        return Response(status_code=204)

    return r

File: tools/agent-tries-baml/convex/workers.ts
// Presence-only table: every long-lived processor heartbeats here so the
// dashboard can show a live roster of agents. NOT a claimable queue — the
// workers table has no queueFields, so it must never be wired into the
// generic queue verbs (claim/transition would 500).

import { query, mutation } from "./_generated/server";
import { v } from "convex/values";
import * as lib from "./lib";

const T = "workers" as const;

/**
 * Fetch one worker presence row by id.
 *
 * @param id - Worker document id to look up.
 * @returns The worker document, or null if it doesn't exist.
 */
export const get = query({
  args: { id: v.string() },
  handler: (ctx, a) => lib.getDoc(ctx, T, a.id),
});

/**
 * List worker presence rows, optionally filtered by role.
 *
 * @param role - Optional role to filter on (scans by_role_status).
 * @param limit - Maximum number of rows to return (defaults to 100).
 * @returns Up to `limit` worker documents.
 */
export const list = query({
  args: { role: v.optional(v.string()), limit: v.optional(v.number()) },
  handler: async (ctx, a) => {
    const limit = a.limit ?? 100;
    if (a.role) {
      return await ctx.db
        .query("workers")
        .withIndex("by_role_status", (q) => q.eq("role", a.role!))
        .take(limit);
    }
    return await ctx.db.query("workers").order("desc").take(limit);
  },
});

/**
 * Upsert a worker's presence heartbeat, keyed by workerId.
 *
 * Patches the existing row when the worker is known, inserts otherwise.
 * Always stamps lastHeartbeat to now.
 *
 * @param workerId - Stable processor identity (role-host-pid-hex).
 * @param role - Processor role (baml_worker, baml_dedup, changelog_worker, ...).
 * @param status - "idle" or "busy".
 * @param currentItemId - The row the worker is processing, when busy.
 * @returns The upserted document id.
 */
export const upsert = mutation({
  args: {
    workerId: v.string(),
    role: v.string(),
    status: v.string(),
    currentItemId: v.optional(v.string()),
  },
  handler: async (ctx, a) => {
    const now = Date.now();
    const existing = await ctx.db
      .query("workers")
      .withIndex("by_worker", (q) => q.eq("workerId", a.workerId))
      .first();
    if (existing) {
      await ctx.db.patch(existing._id, {
        role: a.role,
        status: a.status,
        currentItemId: a.status === "busy" ? a.currentItemId : undefined,
        lastHeartbeat: now,
      });
      return existing._id;
    }
    return await ctx.db.insert("workers", {
      workerId: a.workerId,
      role: a.role,
      status: a.status,
      currentItemId: a.status === "busy" ? a.currentItemId : undefined,
      lastHeartbeat: now,
    });
  },
});
